lib/mongo/crypt/encryption_io.rb

```diff
@@ -95,29 +101,68 @@ def mark_command(cmd)
       def feed_kms(kms_context)
-        endpoint = kms_context.endpoint
-        message = kms_context.message
+        with_ssl_socket(kms_context.endpoint) do |ssl_socket|
 
-        # There is no specific timeout written in the spec. See SPEC-1394
-        # for a discussion and updates on what this timeout should be.
-        socket_timeout = 10
+          Timeout.timeout(SOCKET_TIMEOUT, Error::SocketTimeoutError,
+            'Socket write operation timed out'
+          ) do
+            ssl_socket.syswrite(kms_context.message)
+          end
 
-        host, port = endpoint.split(':')
-        port ||= 443
+          bytes_needed = kms_context.bytes_needed
+          while bytes_needed > 0 do
+            bytes = Timeout.timeout(SOCKET_TIMEOUT, Error::SocketTimeoutError,
+              'Socket read operation timed out'
+            ) do
+              ssl_socket.sysread(bytes_needed)
+            end
 
-        ssl_socket = Socket::SSL.new(host, port, host, socket_timeout, Socket::PF_INET)
-        ssl_socket.write(message)
+            kms_context.feed(bytes)
+            bytes_needed = kms_context.bytes_needed
+          end
+        end
+      end
 
-        num_bytes_needed = kms_context.bytes_needed
+      private
 
-        while num_bytes_needed > 0
-          bytes = []
-          while !ssl_socket.eof?
-            bytes << ssl_socket.readbyte
+      # Provide an SSL socket to be used for KMS calls in a block API
+      #
+      # @param [ String ] endpoint The URI at which to connect the SSL socket
+      # @param [ Proc ] block The block to execute
+      #
+      # @raise [ Mongo::Error::KmsError ] If the socket times out or raises
+      #   an exception
+      #
+      # @note The socket is always closed when the provided block has finished
+      #   executing
+      def with_ssl_socket(endpoint)
+        host, port = endpoint.split(':')
+        port ||= 443 # Default port for AWS KMS API
+
+        begin
+          # Create TCPSocket and set nodelay option
+          tcp_socket = TCPSocket.open(host, port)
+          tcp_socket.setsockopt(::Socket::IPPROTO_TCP, ::Socket::TCP_NODELAY, 1)
+
+          ssl_socket = OpenSSL::SSL::SSLSocket.new(tcp_socket)
+          ssl_socket.sync_close = true # tcp_socket will be closed when ssl_socket is closed
+          ssl_socket.hostname = "#{host}:#{port}" # perform SNI
+
+          Timeout.timeout(
+            SOCKET_TIMEOUT,
+            Error::SocketTimeoutError,
+            'Socket connection timed out'
+          ) do
+            ssl_socket.connect
           end
 
-          kms_context.feed(bytes.pack('C*'))
-          num_bytes_needed = kms_context.bytes_needed
+          yield(ssl_socket)
+        rescue => e
+          raise Error::KmsError, "Error decrypting data key. #{e.class}: #{e.message}"
+        ensure
+          # If there is an error during socket creation, the
+          # ssl_socket object won't exist in this scope
+          ssl_socket.sysclose if ssl_socket
```

Comment:
This close could conceivably also fail, especially since it is an ssl socket and close may perform i/o. Either putting the close call under the rescue raising KmsError above or adding `rescue nil` should be reasonable measures to deal with this. I don't know if the close call should also be guarded with a timeout; I suppose it won't hurt if you wish to do this.